include/deal.II/base/linear_index_iterator.h

```diff
@@ -0,0 +1,449 @@
+   * container.
+   */
+  bool operator <= (const DerivedIterator &) const;
+
+  /**
+   * Comparison operator: uses the same ordering as operator>(), but also
+   * checks for equality.
+   *
+   * This function is only valid if both iterators point into the same
+   * container.
+   */
+  bool operator >= (const DerivedIterator &) const;
+
+  /**
+   * Comparison operator. Result is true if either the first row number is
+   * smaller or if the row numbers are equal and the first index is smaller.
+   *
+   * This function is only valid if both iterators point into the same
+   * container.
+   */
+  bool operator < (const DerivedIterator &) const;
+
+  /**
+   * Comparison operator. Works in the same way as operator<(), just the other
+   * way round.
+   */
+  bool operator > (const DerivedIterator &) const;
+
+protected:
+  /**
+   * Constructor with linear indexing. A default constructor should be
+   * implemented by the inheriting class.
+   */
+  LinearIndexIterator () = default;
+
+  /**
+   * Constructor that copies an accessor.
+   */
+  LinearIndexIterator (const AccessorType accessor);
+
+protected:
+  /**
+   * Store an object of the accessor class.
+   */
+  AccessorType accessor;
+};
+
+
+
+template <class DerivedIterator, class AccessorType>
+inline
+DerivedIterator &
+LinearIndexIterator<DerivedIterator, AccessorType>::operator = (const DerivedIterator &it)
+{
+  accessor.object = it.object;
+  accessor.linear_index = it.linear_index;
+  return static_cast<DerivedIterator &>(*this);
+}
+
+
+
+template <class DerivedIterator, class AccessorType>
+inline
+DerivedIterator &
+LinearIndexIterator<DerivedIterator, AccessorType>::operator ++ ()
+{
+  ++accessor.linear_index;
+  return static_cast<DerivedIterator &>(*this);
+}
+
+
+
+template <class DerivedIterator, class AccessorType>
+inline
+DerivedIterator
+LinearIndexIterator<DerivedIterator, AccessorType>::operator ++ (int)
+{
+  const DerivedIterator copy(this->accessor);
+  ++(*this);
+  return copy;
+}
+
+
+
+template <class DerivedIterator, class AccessorType>
+inline
+DerivedIterator &
+LinearIndexIterator<DerivedIterator, AccessorType>::operator -- ()
+{
+  --accessor.linear_index;
+  return static_cast<DerivedIterator &>(*this);
+}
+
+
+
+template <class DerivedIterator, class AccessorType>
+inline
+DerivedIterator
+LinearIndexIterator<DerivedIterator, AccessorType>::operator -- (int)
+{
+  const DerivedIterator copy(this->accessor);
+  --(*this);
+  return copy;
+}
+
+
+
+template <class DerivedIterator, class AccessorType>
+DerivedIterator
+LinearIndexIterator<DerivedIterator, AccessorType>::
+operator + (const difference_type n) const
+{
+  DerivedIterator copy(this->accessor);
+  copy += n;
+  return copy;
+}
+
+
+
+template <class DerivedIterator, class AccessorType>
+DerivedIterator
+LinearIndexIterator<DerivedIterator, AccessorType>::
+operator - (const difference_type n) const
+{
+  DerivedIterator copy(this->accessor);
```

Comment:
same here

Reply:
Same remark: I don't understand what you meant, since `copy` needs to be mutable.